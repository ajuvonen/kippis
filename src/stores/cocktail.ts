import {
  getByIngredients,
  getBySearchString,
  getDetails,
  getNonAlcoholic,
  getByFirstLetter,
} from '@/api';
import {SEARCHABLE_ALCOHOLS} from '@/utils/constants';
import type {FullDetailsCocktail, SearchResultCocktail, SearchStringProps} from '@/utils/types';
import {defineStore} from 'pinia';
import {compose, flatten, map, prop, sortBy, uniq, without} from 'ramda';

export const useCocktailStore = defineStore('cocktail', {
  state: () => ({
    currentSearch: {} as SearchStringProps,
    cocktailDetails: [] as FullDetailsCocktail[],
    preventFetch: false,
    searchResults: [] as SearchResultCocktail[],
    selection: new Set<number>(),
    highlightedCocktail: null as number | null,
  }),
  getters: {
    getCocktailDetails: (state) => (id: number) => state.cocktailDetails.find((cocktail) => cocktail.id === id) || null,
    // Get a "shopping list" of ingredients. They should be unique, sorted and without some obvious items
    getAllIngredients(state): string[] {
      return compose(
        sortBy(prop(0)),
        without(['water', 'ice']),
        uniq,
        flatten<string[][]>,
        map((id: number) =>
          // Make cream, egg, and citrus fruits show up as their base form 
          this.getCocktailDetails(id)!.ingredients.map(({ingredient}) =>
            ingredient.replace(/^whipped | peel$| spiral$| white$| yolk$/, ''),
          ),
        ),
      )(Array.from(state.selection));
    },
  },
  actions: {
    async addToSelection(id: number) {
      await this.fetchCocktail(id);
      this.selection.add(id);
    },
    async fetchCocktail(id: number) {
      if (!this.cocktailDetails.some((cocktail) => cocktail.id === id)) {
        this.cocktailDetails.push(await getDetails(id));
      }
    },
    async openCocktailModal(id: number) {
      await this.fetchCocktail(id);
      this.highlightedCocktail = id;
    },
    removeFromSelection(id: number) {
      this.selection.delete(id);
    },
    async search(searchString: string) {
      if (searchString.length === 1) {
        this.searchResults = await getByFirstLetter(searchString);
      } else {
        this.searchResults = await getBySearchString(searchString);
      }
    },
    async searchNonAlcoholic() {
      this.searchResults = await getNonAlcoholic();
    },
    async searchWithTag(tag: string) {
      this.searchResults = await getByIngredients(
        SEARCHABLE_ALCOHOLS.find((alcohol) => tag === alcohol.tag)?.ingredients!,
      );
    },
  },
});
