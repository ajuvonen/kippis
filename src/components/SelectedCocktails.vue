<script setup lang="ts">
import {useRoute} from 'vue-router';
import {storeToRefs} from 'pinia';
import {useI18n} from 'vue-i18n';
import {useCocktailStore} from '@/stores/cocktail';
import LinkButton from '@/components/LinkButton.vue';
import SearchResults from '@/components/SearchResults.vue';

const route = useRoute();
const {t} = useI18n();

const cocktailStore = useCocktailStore();
const {selection} = storeToRefs(cocktailStore);
</script>
<template>
  <Transition>
    <aside v-if="selection.length" class="w-[400px] hidden md:flex flex-col bg-slate-800">
      <h2>{{ t('selectedCocktails.title') }}</h2>
      <SearchResults class="pt-2 mb-4 overflow-y-scroll" :cocktails="selection" />
      <LinkButton v-if="route.name === 'search'" to="/ingredients" class="flex-shrink-0" data-test-id="selected-cocktails__action-button">
        {{ t('selectedCocktails.readyButton') }}
      </LinkButton>
      <LinkButton v-else to="/search" class="flex-shrink-0" data-test-id="selected-cocktails__action-button">
        {{ t('selectedCocktails.backButton') }}
      </LinkButton>
    </aside>
  </Transition>
</template>
<style lang="scss" scoped>
.v-enter-to,
.v-leave-from {
  margin-left: 0;
}

.v-enter-active,
.v-leave-active {
  transition: margin 0.2s;
}

.v-enter-from,
.v-leave-to {
  margin-left: -400px;
}
</style>
