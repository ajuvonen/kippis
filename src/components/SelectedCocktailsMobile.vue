<script setup lang="ts">
import {ref, watch} from 'vue';
import {storeToRefs} from 'pinia';
import {useCocktailStore} from '@/stores/cocktail';
import IconComponent from '@/components/IconComponent.vue';
import SelectedCocktailsContent from '@/components/SelectedCocktailsContent.vue';

const cocktailStore = useCocktailStore();
const {selection} = storeToRefs(cocktailStore);

const selectedCocktailsOpen = ref(false);

watch(selection, () => {
  if (!selection.value.length) {
    selectedCocktailsOpen.value = false;
  }
});
</script>
<template>
  <Transition
    enter-from-class="ml-[-100%]"
    enter-to-class="ml-0"
    leave-from-class="ml-0"
    leave-to-class="ml-[-100%]"
    enter-active-class="transition-all"
    leave-active-class="transition-all"
  >
    <button
      v-if="selection.length"
      class="selected-cocktails__mobile-button"
      :class="
        selectedCocktailsOpen
          ? 'selected-cocktails__mobile-button--top'
          : 'selected-cocktails__mobile-button--bottom'
      "
      :aria-label="
        $t(
          selectedCocktailsOpen
            ? 'selectedCocktails.closeSelectedCocktails'
            : 'selectedCocktails.openSelectedCocktails',
        )
      "
      @click="selectedCocktailsOpen = !selectedCocktailsOpen"
    >
      <IconComponent :icon="selectedCocktailsOpen ? 'down' : 'cocktail'" />
    </button>
  </Transition>
  <Transition
    enter-from-class="top-[100%]"
    enter-to-class="top-0"
    leave-from-class="top-0"
    leave-to-class="top-[100%]"
    enter-active-class="transition-all"
    leave-active-class="transition-all"
  >
    <aside v-if="selectedCocktailsOpen && selection.length" class="selected-cocktails pt-10">
      <SelectedCocktailsContent />
    </aside>
  </Transition>
</template>
<style lang="scss" scoped>
.selected-cocktails__mobile-button {
  @apply absolute z-20 left-0 w-full mb-0 py-2 rounded-none border-l-0 border-r-0;
}

.selected-cocktails__mobile-button--top {
  @apply top-0 rounded-bl-md rounded-br-md border-t-0;
}

.selected-cocktails__mobile-button--bottom {
  @apply bottom-0 rounded-tl-md rounded-tr-md border-b-0;
}

.selected-cocktails {
  @apply absolute z-10 w-full h-full flex flex-col bg-slate-800;
}

@media print {
  .selected-cocktails {
    @apply hidden;
  }
}
</style>
