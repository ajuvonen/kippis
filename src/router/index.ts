import {createRouter, createWebHistory} from 'vue-router';
import {isEmpty} from 'remeda';
import HomeView from '@/views/HomeView.vue';
import {useCocktailStore} from '@/stores/cocktail';

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      path: '/',
      name: 'home',
      component: HomeView,
    },
    {
      path: '/search',
      name: 'search',
      props: (route) => ({
        searchString: !route.query.tag ? route.query.searchString : undefined,
        tag: route.query.tag,
      }),
      beforeEnter: (to) => {
        if (isEmpty(to.query)) {
          const cocktailStore = useCocktailStore();
          if (cocktailStore.currentSearch.tag || cocktailStore.currentSearch.searchString) {
            cocktailStore.preventFetch = true;
            return {path: to.path, query: {...cocktailStore.currentSearch}};
          }
        }
      },
      component: () => import('@/views/SearchView.vue'),
    },
    {
      path: '/ingredients',
      name: 'ingredients',
      component: () => import('@/views/IngredientsView.vue'),
    },
  ],
});

export default router;
