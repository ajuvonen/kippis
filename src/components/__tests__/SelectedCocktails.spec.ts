import {describe, it, expect, beforeEach} from 'vitest';
import {mount} from '@vue/test-utils';
import router from '@/router';
import {useCocktailStore} from '@/stores/cocktail';
import {testCocktails} from '@/components/__tests__/mswHandlers';
import SelectedCocktails from '@/components/SelectedCocktails.vue';
import SelectedCocktailsMobile from '@/components/SelectedCocktailsMobile.vue';

describe('SelectedCocktails', () => {
  let cocktailStore: ReturnType<typeof useCocktailStore>;

  beforeEach(async () => {
    cocktailStore = useCocktailStore();
    await router.push('/search');
  });

  it('mounts', () => {
    cocktailStore.selection.push(testCocktails[0][1]);
    const wrapper = mount(SelectedCocktails);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('prints selection data', () => {
    cocktailStore.selection = [testCocktails[0][1], testCocktails[1][1]];
    const wrapper = mount(SelectedCocktails);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('changes button text depending of page', async () => {
    cocktailStore.selection.push(testCocktails[1][1]);
    const wrapper = mount(SelectedCocktails);
    expect(wrapper.findByTestId('selected-cocktails__action-button').text()).toBe(
      'Ready? Proceed to instructions.',
    );
    await router.push('/instructions');
    expect(wrapper.findByTestId('selected-cocktails__action-button').text()).toBe(
      'Thirsty? Return to search.',
    );
  });
});

describe('SelectedCocktailsMobile', () => {
  let cocktailStore: ReturnType<typeof useCocktailStore>;

  beforeEach(async () => {
    cocktailStore = useCocktailStore();
    await router.push('/search');
  });

  it('mounts', () => {
    cocktailStore.selection.push(testCocktails[0][1]);
    const wrapper = mount(SelectedCocktailsMobile);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('prints selection data', async () => {
    cocktailStore.selection = [testCocktails[0][1], testCocktails[1][1]];
    const wrapper = mount(SelectedCocktailsMobile);
    await wrapper.find('.selected-cocktails__mobile-button').trigger('click');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
