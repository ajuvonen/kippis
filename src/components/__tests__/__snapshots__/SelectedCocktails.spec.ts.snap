// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SelectedCocktails > mounts 1`] = `
"<transition-stub data-v-418939cb=\\"\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
  <aside data-v-418939cb=\\"\\" class=\\"w-[400px] hidden md:flex flex-col bg-slate-800\\">
    <h2 data-v-418939cb=\\"\\">Your Cocktails</h2>
    <ul data-v-418939cb=\\"\\" class=\\"flex flex-auto flex-wrap gap-6 content-start justify-center pt-2 mb-4 overflow-y-scroll\\">
      <li>
        <div data-v-cd13f5f7=\\"\\" class=\\"search-result__wrapper\\">
          <div data-v-cd13f5f7=\\"\\" aria-label=\\"View info about &quot;Margarita&quot;\\" class=\\"search-result__image-wrapper\\" role=\\"button\\" tabindex=\\"0\\">
            <div data-v-c27c5544=\\"\\" class=\\"lazy-image__fallback\\"><svg data-v-c27c5544=\\"\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" class=\\"w-5 h-5\\">
                <path d=\\"M7.5,7L5.5,5H18.5L16.5,7M11,13V19H6V21H18V19H13V13L21,5V3H3V5L11,13Z\\" class=\\"transition-all ease-in-out\\"></path>
              </svg></div>
            <div data-v-cd13f5f7=\\"\\" class=\\"search-result__shadow\\">Margarita</div>
          </div><button data-v-634d56d3=\\"\\" data-v-cd13f5f7=\\"\\" aria-label=\\"Remove &quot;Margarita&quot; from selection\\" class=\\"remove-cocktail-0 action-button bg-rose-400 border-slate-800\\"><svg data-v-634d56d3=\\"\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" class=\\"w-5 h-5\\">
              <path d=\\"M9,3V4H4V6H5V19A2,2 0 0,0 7,21H17A2,2 0 0,0 19,19V6H20V4H15V3H9M9,8H11V17H9V8M13,8H15V17H13V8Z\\" class=\\"transition-all ease-in-out\\"></path>
            </svg></button>
        </div>
      </li>
    </ul><a data-v-418939cb=\\"\\" href=\\"/ingredients\\" class=\\"flex items-center link-button flex-shrink-0\\" data-test-id=\\"selected-cocktails__action-button\\">Ready? Proceed to ingredients.</a>
  </aside>
</transition-stub>"
`;

exports[`SelectedCocktails > prints selection data 1`] = `
"<transition-stub data-v-418939cb=\\"\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
  <aside data-v-418939cb=\\"\\" class=\\"w-[400px] hidden md:flex flex-col bg-slate-800\\">
    <h2 data-v-418939cb=\\"\\">Your Cocktails</h2>
    <ul data-v-418939cb=\\"\\" class=\\"flex flex-auto flex-wrap gap-6 content-start justify-center pt-2 mb-4 overflow-y-scroll\\">
      <li>
        <div data-v-cd13f5f7=\\"\\" class=\\"search-result__wrapper\\">
          <div data-v-cd13f5f7=\\"\\" aria-label=\\"View info about &quot;Margarita&quot;\\" class=\\"search-result__image-wrapper\\" role=\\"button\\" tabindex=\\"0\\">
            <div data-v-c27c5544=\\"\\" class=\\"lazy-image__fallback\\"><svg data-v-c27c5544=\\"\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" class=\\"w-5 h-5\\">
                <path d=\\"M7.5,7L5.5,5H18.5L16.5,7M11,13V19H6V21H18V19H13V13L21,5V3H3V5L11,13Z\\" class=\\"transition-all ease-in-out\\"></path>
              </svg></div>
            <div data-v-cd13f5f7=\\"\\" class=\\"search-result__shadow\\">Margarita</div>
          </div><button data-v-634d56d3=\\"\\" data-v-cd13f5f7=\\"\\" aria-label=\\"Remove &quot;Margarita&quot; from selection\\" class=\\"remove-cocktail-0 action-button bg-rose-400 border-slate-800\\"><svg data-v-634d56d3=\\"\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" class=\\"w-5 h-5\\">
              <path d=\\"M9,3V4H4V6H5V19A2,2 0 0,0 7,21H17A2,2 0 0,0 19,19V6H20V4H15V3H9M9,8H11V17H9V8M13,8H15V17H13V8Z\\" class=\\"transition-all ease-in-out\\"></path>
            </svg></button>
        </div>
      </li>
      <li>
        <div data-v-cd13f5f7=\\"\\" class=\\"search-result__wrapper\\">
          <div data-v-cd13f5f7=\\"\\" aria-label=\\"View info about &quot;Mojito&quot;\\" class=\\"search-result__image-wrapper\\" role=\\"button\\" tabindex=\\"0\\">
            <div data-v-c27c5544=\\"\\" class=\\"lazy-image__fallback\\"><svg data-v-c27c5544=\\"\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" class=\\"w-5 h-5\\">
                <path d=\\"M7.5,7L5.5,5H18.5L16.5,7M11,13V19H6V21H18V19H13V13L21,5V3H3V5L11,13Z\\" class=\\"transition-all ease-in-out\\"></path>
              </svg></div>
            <div data-v-cd13f5f7=\\"\\" class=\\"search-result__shadow\\">Mojito</div>
          </div><button data-v-634d56d3=\\"\\" data-v-cd13f5f7=\\"\\" aria-label=\\"Remove &quot;Mojito&quot; from selection\\" class=\\"remove-cocktail-1 action-button bg-rose-400 border-slate-800\\"><svg data-v-634d56d3=\\"\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" class=\\"w-5 h-5\\">
              <path d=\\"M9,3V4H4V6H5V19A2,2 0 0,0 7,21H17A2,2 0 0,0 19,19V6H20V4H15V3H9M9,8H11V17H9V8M13,8H15V17H13V8Z\\" class=\\"transition-all ease-in-out\\"></path>
            </svg></button>
        </div>
      </li>
    </ul><a data-v-418939cb=\\"\\" href=\\"/ingredients\\" class=\\"flex items-center link-button flex-shrink-0\\" data-test-id=\\"selected-cocktails__action-button\\">Ready? Proceed to ingredients.</a>
  </aside>
</transition-stub>"
`;
